Poll for devtools server, with notification

johnjbarton opened the pull request:
A more ambitious and of course risky approach: poll the server and attempt to sync the existence of the devtools UI to the server being healthy.

cjihrig commented:
Thanks for the PR. I've noticed a few things.

1. The changes to the settings page break the save functionality on my machine. Possibly the window closes before the saves actually go through. These changes should probably be a separate PR anyway.
2. The polling is too aggressive to be usable for me personally. Closing DevTools while Node is still running causes new windows to keep relaunching.
3. I'm not sure the notification system is really working correctly.
4. I don't necessarily want DevTools to exit completely when Node exits. Sometimes I leave DevTools running after Node exits to, for example, analyze snapshots or profiles.

What I would prefer is:
1. Polling be an option that can be turned on or off on the settings page.
2. Polling starts once the user tries to connect via the Chrome extensions. Polling should not go on forever in the background when the extension is installed.
2. When polling is turned on, DevTools windows are only closed when Node starts or restarts. This would make sure DevTools is left open at the end of the session.
3. If the user explicitly closes out of DevTools, polling stops.

johnjbarton commented:
Thanks for the review! I'll need some more information to re-try.

1. Yes, save was intended to a be a separate PR, sorry.

2. "The polling is too aggressive to be usable for me personally. Closing DevTools while Node is still running causes new windows to keep relaunching."  
  Can you help me understand why you are closing devtools while node is still running under the debugger?  What is your goal in that case? (I expect users to be focused on the command line).
  When you say "a causes new windows to keep relaunching" do you mean: "I meant to close devtools but it reopened" or "more than one window opens at a time"?

3. "I'm not sure the notification system is really working correctly." Can you give me any hints about the problems you see? It works great for me so I don't know what to fix.

4."I don't necessarily want DevTools to exit completely when Node exits. Sometimes I leave DevTools running after Node exits to, for example, analyze snapshots or profiles." 

For me, when my node app exits, node puts up a line on the command line "Waiting for the debugger to disconnnect..." This allows time to analyze snapshots, set break points etc. After Node completely exits, Devtools puts up a non-responsive "reconnect" button, covers the screen, and prevents further interaction. 

Is this not what you see?

I am using Chrome 55,  node v 6.9, on ubuntu linux. Maybe your setup different in some dimension that I need to factor in?

1. Polling be an option that can be turned on or off on the settings page.
I agree.

2. "Polling starts once the user tries to connect via the Chrome extensions. Polling should not go on forever in the background when the extension is installed."

This is already how the PR works: polling starts when the user clicks the icon and stops when the user clicks the notification. The polling does not continue forever.  Maybe your issue with the notification is related?

3. "When polling is turned on, DevTools windows are only closed when Node starts or restarts. This would make sure DevTools is left open at the end of the session." 

Sorry I completely do not understand how this could work. If devtools windows are closed when node starts, there is no UI for debugging? And devtools is already left open at the end of the session, so I'm confused by your request.

4. "If the user explicitly closes out of DevTools, polling stops."  
Do you mean: "If the user explicitly closes out of Devtools, devtools should not reopen until the user restarts debugging"? I think that is reasonable and feasible. But stopping the polling if the user closes the devtools somewhat defeats the user experience: having one simple repeatable UI to go through the debug loop (control + C on the command line, up arrow, return; that is as good as we can do for now to be equal to reloading a web page).  But I will see if this is feasible (UI on the devtools page is not part of the current PR).

johnjbarton committed:
Poll for devtools server, with notification.

    Add option 'poll' to the settings UI.

    When 'poll' is true, and the extension icon is clicked, enter a polling loop:
      If the devtools json URL responds, open inspector and pull down notification.
      Else put up the notification and re-poll.

    Everytime the devtools server comes up, devtools opens;
    Everytime the devtools server exits, devtools closes.

    On the command line, run nodejs with debug flags, devtools server comes up,
    devtools opens in the browser.
    Kill nodejs, devtools closes in the browser.

    Whenever devtools is closed, notification "Polling for debug server,
    dismiss to stop polling" is shown.  This is the UI to exit polling.

    If the devtools UI tab is closed by the user, exit polling.

    Not yet as good as control+R in web debugging, but on the track.

cjihrig commented:
This has conflicts since landing #30. Can you rebase, squash, and add a checkbox to the settings page (off by default for now) so that polling can be turned on and off. After that, I think I'm comfortable landing it. Then, we can iterate if there are issues. Thanks again for this.

johnjbarton commented:
I've tried to address the user-experience issues you raised. Try this out and let me knos.

cjihrig commented:
This is really good. I like it a lot. Thanks for implementing it and working through my feedback. Landed in a1132bf261ff9af631339710889f07465eef5aa8 with fixes for a couple edge cases.

Note that the case where you use the extension via the UI ("Quick launch using default settings" is unchecked) and polling mode is enabled still has a few hiccups.

johnjbarton commented:
> Note that the case where you use the extension via the UI ("Quick launch using default settings" is unchecked) and polling mode is enabled still has a few hiccups.

Oh sorry I forgot to check this case altogether!  I'll try it and see what can be done.

cjihrig commented:
Don't worry about it. I took care of it in aebaa8c1192b860db9054f2cd92e3fbc760c694b. Just published a new version with your changes. Just waiting for the Chrome store to update. Thanks again!
